### PR TITLE
[WIP] XDG support

### PR DESCRIPTION
**Pull request by Yamakaky**
This PR is meant to bring XDG support to neovim. Features : 
- Expand XDG_\* variables
- Change default files location to match the XDG convention.

The [tmp] commit will be squashed before merge.

### CODE REVIEW

File: src/nvim/os/env.c [hunk 1]
```diff
@@ -427,6 +427,31 @@ char *vim_getenv(const char *name)
...
+      home_suffix = "/.local/share/";
+    } else if (strcmp(name, "XDG_CONFIG_HOME") == 0) {
+      home_suffix = "/.config/";
+    } else if (strcmp(name, "XDG_CACHE_HOME") == 0) {
+      home_suffix = "/.cache/";
+    } else if (strcmp(name, "XDG_RUNTIME_DIR") == 0) {
+      return xstrdup("/tmp/");  // Arbitrary value, any idea?
+    }
+
+    // We need to apprend the prefix to the HOME path.
```
Comment (fmoralesc) [inline src/nvim/os/env.c]: apprend -> append

File: src/nvim/os/env.c [hunk 2]
```diff
@@ -427,6 +427,31 @@ char *vim_getenv(const char *name)
...
+
+    const char *home_suffix = "";
+    if (strcmp(name, "XDG_DATA_HOME") == 0) {
+      home_suffix = "/.local/share/";
+    } else if (strcmp(name, "XDG_CONFIG_HOME") == 0) {
+      home_suffix = "/.config/";
+    } else if (strcmp(name, "XDG_CACHE_HOME") == 0) {
+      home_suffix = "/.cache/";
+    } else if (strcmp(name, "XDG_RUNTIME_DIR") == 0) {
+      return xstrdup("/tmp/");  // Arbitrary value, any idea?
```
Comment (ZyX-I) [inline src/nvim/os/env.c]: I am using `~/.run`. Note that you will have to create directory if it does not exist, and it has specific permission requirements (i.e. 0700: u=rw, group and everybody else have no permissions).

File: src/nvim/os/env.c [hunk 3]
```diff
@@ -427,6 +427,31 @@ char *vim_getenv(const char *name)
...
+    }
+
+    // We need to apprend the prefix to the HOME path.
+    char *value = vim_getenv("HOME");
+    value = xrealloc(value, strlen(value) + strlen(home_suffix) + 1);
+    strcat(value, home_suffix);
```
Comment (fmoralesc) [inline src/nvim/os/env.c]: clint is warning about this strcat()

Comment (Yamakaky) [inline src/nvim/os/env.c]: Saw that, on it.

Comment (ZyX-I) [inline src/nvim/os/env.c]: If you already have to compute length better use

    const size_t value_len = strlen(value);
    const size_t home_suffix_len = strlen(home_suffix);
    value = xrealloc(value, value_len + home_suffix_len + 1);
    memmove(value + value_len, home_suffix, home_suffix_len + 1);

File: src/nvim/os/env.c [hunk 4]
```diff
@@ -427,6 +427,31 @@ char *vim_getenv(const char *name)
     return xstrdup(kos_env_path);
   }
 
+  // XDG Base Directory expansion.
+  if (strncmp(name, "XDG_", 4) == 0) {
```
Comment (justinmk) [inline src/nvim/os/env.c]: Sneaking this into `vim_getenv` and making it do something totally different than what the caller asked, is unsettling.

File: src/nvim/os/env.c (hunk 2 above)
Comment (ZyX-I) [inline src/nvim/os/env.c]: Another option is `~/.local/run`.

Comment (fmoralesc) [inline src/nvim/os/env.c]: Shouldn't the `/run/...` directory be used for stuff like this? I'm not sure how standard that is, though.

Comment (ZyX-I) [inline src/nvim/os/env.c]: @fmoralesc `/run` is pretty standard. But I do not know any standard option for user-specific `...` (AFAIR there are some distribution-specific, but I do not remember where I saw this and do not find anything like this in my `/run`), so am using `~/…`.

Comment (fmoralesc) [inline src/nvim/os/env.c]: No, I meant a directory inside /run, not something called "..." inside it. Sorry for the confusion.

Comment (ZyX-I) [inline src/nvim/os/env.c]: @fmoralesc I did not mean anything named *literally* `...`. I said that I do not know what it may be replaced with.

File: src/nvim/os/env.c (hunk 4 above)
Comment (Yamakaky) [inline src/nvim/os/env.c]: As it already expand $HOME and others, I thought it was appropriate, but I don't mind changing. I should have updated the doc, though.

Comment (justinmk) [inline src/nvim/os/env.c]: Let's see what others think. I haven't had chance to carefully think about it.

File: src/nvim/os/env.c (hunk 2 above)
Comment (fmoralesc) [inline src/nvim/os/env.c]: Ok. In my system there is a tree for every active user under /run, but I can't confirm the layout atm.

Comment (jck) [inline src/nvim/os/env.c]: systemd (and I think upstart too) automatically creates /run/user/ID with correct permissions and sets XDG_RUNTIME_DIR. $XDG_RUNTIME_DIR is set for me on Archlinux and Ubuntu 14.04.

File: src/nvim/os/env.c (hunk 1 above)
Comment (Yamakaky) [inline src/nvim/os/env.c]: oops ^^

File: src/nvim/os/env.c (hunk 3 above)
Comment (Yamakaky) [inline src/nvim/os/env.c]: Done : https://github.com/neovim/neovim/commit/86e7c300da30755aab4369d0c7e075972361e521

File: src/nvim/os/env.c (hunk 2 above)
Comment (ZyX-I) [inline src/nvim/os/env.c]: @fmoralesc This is AFAIK distribution-specific. I do not have `/run/zyx` (Gentoo, openrc init). Neither I have `/run/user`.

Comment (fmoralesc) [inline src/nvim/os/env.c]: In that case, I think `~/.local/run/` is best.

Comment (Yamakaky) [inline src/nvim/os/env.c]: Same for me, I have /run/usr/<uid> on archlinux/systemd

**Comment by Yamakaky** [comment]
The [tmp] commit will be squashed before merge.

File: src/nvim/os/env.c (hunk 2 above)
Comment (jck) [inline src/nvim/os/env.c]: In any case, it looks like nvim doesn't use XDG_RUNTIME_DIR at all?

Comment (ZyX-I) [inline src/nvim/os/env.c]: I think that XDG_RUNTIME_DIR is a proper default place for NeoVim sockets, much better then `/tmp/nvimXXX/N`.

Comment (ZyX-I) [inline src/nvim/os/env.c]: (Do not forget to call `os_mkdir[_recurse]`: with such default as `~/.local/run` it is needed, and most likely exactly with `_recurse` because sometimes only `~` is present, but not `~/.local`.)

**Comment by fmoralesc** [comment]
I think nvim should still fallback on ~/.nvim/nvimrc and then ~/.nvimrc. How does this handle that?

**Comment by Earnestly** [comment]
Let's just get everyone up to speed on what `XDG_RUNTIME_DIR` is and what the requirements on it are:

* Used for non-essential, user-specific data files such as sockets, named pipes, etc.
* Defaults to nothing, warnings should be issued if not set or equivalents provided.
* Must be owned by the user.
* Access mode must be `0700`.
* Filesystem fully featured by standards of OS.
* Must be on the local filesystem.
* May be subject to periodic cleanup.
* Modified every 6 hours or set sticky bit if persistence is desired.
* Can only exist for the duration of the user's login.
* Shouldn't store large files as it may be mounted as a tmpfs.

Generally you don't create this directory, something like logind would.  However if you do not use logind, you have to create it with the requirements listed above.

**Comment by ZyX-I** [comment]
@fmoralesc Why do you think so? NeoVim is still not stable so it does not need to have backward compatibility with itself. Neither it needs to be backward compatible with Vim on this matter. And making such changes as early as possible creates a good opportunity to throw away some legacy.

I think that making such fallbacks is adding more code for no real reason. It is not much work to move configuration files to the new location.

**Comment by ZyX-I** [comment]
@Earnestly From NeoVim POV only those requirements are essential:

1. 0700, owned by the current user, not shared, allows creating files and directories without any group and other access.
2. Inside FS sockets and directories can be created and created with 0n00 on their own.

First are security requirements, second are functional requirements. Other requirements you listed is XDG specification requirements or common implementation details. When seeking directory to use you can test any requirement from the list, except for the shared status of the directory.

**Comment by jck** [comment]
I think the following should be added to this PR:
1) OSX and windows fallbacks need to be implemented for all XDG vars
2) https://github.com/neovim/neovim/blob/master/src/nvim/msgpack_rpc/server.c#L38 should be modified to create sockets in $XDG_RUNTIME_DIR

**Comment by fmoralesc** [comment]
> Why do you think so? 

@ZyX-I Because otherwise this will break every existing setup on merge. I think the better approach is to add this as an alternative first and remove the fallback later once the change is properly communicated and the changes are tested in the wild. And no, I don't think people read *Following HEAD* in the wiki (I don't think people will read anything, at all).

> NeoVim ... does not need to have backward compatibility with itself.

I disagree, at least in this case.

**Comment by justinmk** [comment]
> I think that making such fallbacks is adding more code for no real reason. It is not much work to move configuration files to the new location.

Yes, we agreed to that in #78. Especially if we get this in before 0.1. 

> I don't think people will read anything, at all).

I agree, but we need to play the alpha card here. Unless the fallback code is trivial, I don't want to spend a lot of time on it. We have *very few* users, really. We have more important things to work on than back-compat for a pre-alpha software. If we get XDG right, I think most will be more than happy with that.

> https://github.com/neovim/neovim/blob/master/src/nvim/msgpack_rpc/server.c#L38 should be modified to create sockets in $XDG_RUNTIME_DIR

:+1:

**Comment by justinmk** [comment]
By they way, XDG support is going to be another strong out-of-the-box improvement, because we can enable backup/undo without dropping files everywhere.

**Comment by Yamakaky** [comment]
@jck I'm not on windows so this PR is for unix only. See win_defs.h

**Comment by fmoralesc** [comment]
> If we get XDG right, I think most will be more than happy with that.

Sure, I just worry some plugins might insta-break with this, and well, I'm a softie.

**Comment by justinmk** [comment]
> Sure, I just worry some plugins might insta-break with this, 

If it turns into a plugin disaster, we can revert and revisit. Ok?

**Comment by fmoralesc** [comment]
In those cases it will probably be the plugin's fault, really, most likely for hardcodding paths. Anyway, I don't think it's terribly likely to have many issues, but plugins that might need to modify the runtimepath (like plugin managers) have a higher chance of failing. If someone could test this, it would be great (I might do it later if none else cares).

**Comment by Yamakaky** [comment]
BTW, I don't see why the travis build fails, the logs looks clean. What's wrong ?

**Comment by fmoralesc** [comment]
Did you try running the functional test locally? It seems to steal at that point

**Comment by jck** [comment]
Regarding XDG_RUNTIME_DIR. From http://wayland.freedesktop.org/building.html

>Weston creates its unix socket file (for example, wayland-0) in the directory specified by the required environment variable $XDG_RUNTIME_DIR. Clients use the same variable to find that socket. This is provided using systemd by some distributions (Fedora, Arch since June 2012 or Exherbo). Ubuntu began providing it in Quantal. It is not provided by Gentoo.

>If you are using a distro that does set $XDG_RUNTIME_DIR for you, you can skip this part. Otherwise, you must set it using your shell profile capability. 
...

**Comment by Earnestly** [comment]
@ ZyX-I Yes, you're correct.  What I posted was mostly just a summary of what it defines from the specification itself.  Of course software need only be concerned with parts of it.

File: src/nvim/os/env.c [hunk 5]
```diff
@@ -427,6 +427,33 @@ char *vim_getenv(const char *name)
...
+    const char *home_suffix = "";
+    if (strcmp(name, "XDG_DATA_HOME") == 0) {
+      home_suffix = "/.local/share/";
+    } else if (strcmp(name, "XDG_CONFIG_HOME") == 0) {
+      home_suffix = "/.config/";
+    } else if (strcmp(name, "XDG_CACHE_HOME") == 0) {
+      home_suffix = "/.cache/";
+    } else if (strcmp(name, "XDG_RUNTIME_DIR") == 0) {
+      return xstrdup("/tmp/");  // Arbitrary value, any idea?
+    }
```
Comment (ZyX-I) [inline src/nvim/os/env.c]: By the way, where are `XDG_{DATA,CONFIG}_DIRS`? One of the suggestions in my issue was initially populating `&runtimepath` from XDG_CONFIG_DIRS+XDG_CONFIG_HOME, but I think that XDG_DATA_{DIRS,HOME} should also be taken into account (reason: /usr/share/vim/vimfiles is where package managers usually put third-party plugins, thus with new directories /usr/share/nvim/site should be included and this is naturally done using XDG_DATA_DIRS, but not XDG_DATA_HOME, XDG_CONFIG_HOME or XDG_CONFIG_DIRS). This way &runtimepath will look like:

    $XDG_CONFIG_HOME/nvim
    $XDG_DATA_HOME/nvim
    $xdg_config_dir/nvim (repeated for each directory in $XDG_CONFIG_DIRS)
    $xdg_data_dir/nvim/site (repeated for each directory in $XDG_DATA_DIRS)
    $VIMRUNTIME (looks something like /usr/share/nvim/runtime)
    $xdg_data_dir/nvim/site/after (repeated for each directory in $XDG_DATA_DIRS)
    $xdg_config_dir/nvim/after (repeated for each directory in $XDG_CONFIG_DIRS)
    $XDG_CONFIG_HOME/nvim/after

`XDG_DATA_DIRS` defaults to `/usr/local/share/:/usr/share/` according to the specification. This is why initial suggestion is using `site` in runtimepath only: if it was `runtime` then $VIMRUNTIME was naturally included (effectively meaning that this setting will become obsolete which is bad for compatibility reasons).

File: src/nvim/os/unix_defs.h [hunk 6]
```diff
@@ -28,46 +32,51 @@
...
 // Default for 'directory'.
 #ifndef DFLT_DIR
-# define DFLT_DIR     ".,~/tmp,/var/tmp,/tmp"
+# define DFLT_DIR     "$XDG_DATA_HOME/nvim/swap//"
 #endif
 
 // Default for 'viewdir'.
 #ifndef DFLT_VDIR
-# define DFLT_VDIR    "~/.nvim/view"
+# define DFLT_VDIR    "$XDG_DATA_HOME/nvim/view/"
```
Comment (ZyX-I) [inline src/nvim/os/unix_defs.h]: Are not views used as configuration? I would use XDG_CONFIG_HOME here. Also this is naturally shortcuted with NVIM_{DATA,CONF}_DIR.

File: src/nvim/os/unix_defs.h [hunk 7]
```diff
@@ -28,46 +32,51 @@
...
 // Default for 'backupdir'.
 #ifndef DFLT_BDIR
-# define DFLT_BDIR    ".,~/tmp,~/"
+# define DFLT_BDIR    ".,$XDG_DATA_HOME/nvim/backup/"
 #endif
 
 // Default for 'directory'.
 #ifndef DFLT_DIR
-# define DFLT_DIR     ".,~/tmp,/var/tmp,/tmp"
+# define DFLT_DIR     "$XDG_DATA_HOME/nvim/swap//"
```
Comment (ZyX-I) [inline src/nvim/os/unix_defs.h]: NVIM_DATA_DIR

File: src/nvim/os/unix_defs.h [hunk 8]
```diff
@@ -28,46 +32,51 @@
...
 #endif
 #ifndef VIMINFO_FILE
-# define VIMINFO_FILE "~/.nviminfo"
+# define VIMINFO_FILE NVIM_DATA_DIR "/viminfo"
 #endif
 
 // Default for 'backupdir'.
 #ifndef DFLT_BDIR
-# define DFLT_BDIR    ".,~/tmp,~/"
+# define DFLT_BDIR    ".,$XDG_DATA_HOME/nvim/backup/"
```
Comment (ZyX-I) [inline src/nvim/os/unix_defs.h]: NVIM_DATA_DIR

File: src/nvim/os/unix_defs.h [hunk 9]
```diff
@@ -19,6 +19,10 @@
 #define SPECIAL_WILDCHAR "`'{"
 
 // Unix system-dependent file names
+
+#define NVIM_CONF_DIR "$XDG_CONFIG_HOME/nvim/"
+#define NVIM_DATA_DIR "$XDG_DATA_HOME/nvim/"
```
Comment (Pyrohh) [inline src/nvim/os/unix_defs.h]: Are you sure the leading slashes should be here? Wouldn't this result in `$XDG_CONFIG_HOME/nvim//...` if something is appended to it?

File: src/nvim/os/env.c [hunk 10]
```diff
@@ -427,6 +427,33 @@ char *vim_getenv(const char *name)
     return xstrdup(kos_env_path);
   }
 
+  // XDG Base Directory expansion.
+  if (strncmp(name, "XDG_", 4) == 0) {
+    const char *home = vim_getenv("HOME");
+    if (!home) {
+        return NULL;
```
Comment (Pyrohh) [inline src/nvim/os/env.c]: indentation is wrong

File: src/nvim/os/env.c (hunk 5 above)
Comment (ZyX-I) [inline src/nvim/os/env.c]: I have added commits that show how in my opinion this should look like: b4331ace2b47260867ef619840fa38331a1f2f41, its parent and parent of its parent. Feel free to pull them (branch `pr-3120` of https://github.com/ZyX_I/neovim) if you like.

File: src/nvim/os/env.c [hunk 11]
```diff
@@ -427,6 +427,33 @@ char *vim_getenv(const char *name)
     return xstrdup(kos_env_path);
   }
 
+  // XDG Base Directory expansion.
```
Comment (Pyrohh) [inline src/nvim/os/env.c]: The function comment above needs to be updated.

File: src/nvim/os/env.c (hunk 5 above)
Comment (ZyX-I) [inline src/nvim/os/env.c]: Err. I forgot about `/site`: 89e1eac51c8bc146ae4131cbaa93e595d4dc8e0f.

**Comment by fmoralesc** [comment]
This needs to create the `$XDG_DATA_HOME/nvim` dir if doesn't exist. I just tried the patches and if errors out in exist because it can''t find where to put the viminfo file.

File: src/nvim/os/env.c (hunk 5 above)
Comment (ZyX-I) [inline src/nvim/os/env.c]: Some reordering: e6c09b101861b2d3cdd59e5c4631ccc175224975.

File: src/nvim/os/unix_defs.h (hunk 6 above)
Comment (Yamakaky) [inline src/nvim/os/unix_defs.h]: Yes, you are right.

File: src/nvim/os/unix_defs.h (hunk 7 above)
Comment (Yamakaky) [inline src/nvim/os/unix_defs.h]: Wow, I forgot to use it for half the variables ^^

File: src/nvim/os/env.c (hunk 10 above)
Comment (Yamakaky) [inline src/nvim/os/env.c]: done

File: src/nvim/os/env.c (hunk 11 above)
Comment (Yamakaky) [inline src/nvim/os/env.c]: done

File: src/nvim/os/unix_defs.h (hunk 9 above)
Comment (Yamakaky) [inline src/nvim/os/unix_defs.h]: It doesn't matter, test with `cd /usr////////bin`

File: src/nvim/os/env.c (hunk 2 above)
Comment (Yamakaky) [inline src/nvim/os/env.c]: nvim doesn't use XDG_RUNTIME_DIR, but it should. I don't know what to modify, though. I think /tmp is a good default. See http://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html#variables : it MUST NOT survive a reboot.

File: src/nvim/os/env.c (hunk 5 above)
Comment (Yamakaky) [inline src/nvim/os/env.c]: You are right. It's just that it's a bigger change than simply using *_HOME, since you can't expand XDG_*_DIR in macros, it needs code changes.

**Comment by Yamakaky** [comment]
Where in the code should I add this ?

File: src/nvim/os/env.c (hunk 2 above)
Comment (ZyX-I) [inline src/nvim/os/env.c]: @Yamakaky You have *completely* wrong priorities. `/tmp` is not 0700 and is not owned by the current user and *that* should have more priority.

Note: XDG_RUNTIME_DIR according to the spec is not something that should be used like $XDG_RUNTIME_DIR/$USER/…, it is $XDG_RUNTIME_DIR/….

**Comment by fmoralesc** [comment]
I reviewed the most common plugin managers, and I found:

* pathogen shouldn't have any problems with this
* vim-plug shouldn't either
* Vundle shouldn't either
* VAM assumes '~/vim': https://github.com/MarcWeber/vim-addon-manager/blob/master/autoload/vam.vim#L36
* NeoBundle also assumes it as a default: https://github.com/Shougo/neobundle.vim/blob/master/autoload/neobundle.vim#L179

**Comment by ZyX-I** [comment]
> Where in the code should I add this ?

I would suggest to use `write_viminfo` function from ex_cmds.c. Though you may wait until #2506: this way you will not have to add any code because I already create directory there. This PR has only tests left, but a lot of them.

File: src/nvim/os/env.c [hunk 12]
```diff
@@ -427,6 +428,37 @@ char *vim_getenv(const char *name)
...
+    } else if (strcmp(name, "XDG_RUNTIME_DIR") == 0) {
+      return xstrdup("/tmp/");  // Arbitrary value, any idea?
+    } else if (strcmp(name, "XDG_CONFIG_DIRS") == 0) {
+      return xstrdup("/etc/xdg");
+    } else if (strcmp(name, "XDG_DATA_DIRS") == 0) {
+      return xstrdup("/usr/local/share/:/usr/share/");
+    }
+
+    // We need to append the prefix to the HOME path.
+    char *value = vim_getenv("HOME");
```
Comment (ZyX-I) [inline src/nvim/os/env.c]: You already have `home`.

File: src/nvim/os/env.c [hunk 13]
```diff
@@ -427,6 +428,37 @@ char *vim_getenv(const char *name)
...
+    const size_t value_len = strlen(value);
+    const size_t home_suffix_len = strlen(home_suffix);
+    value = xrealloc(value, value_len + home_suffix_len + 1);
+    memmove(value + value_len, home_suffix, home_suffix_len + 1);
+    return value;
```
Comment (ZyX-I) [inline src/nvim/os/env.c]: Memory leak: `vim_getenv` returns an allocated value. `home` is neither used nor freed.

**Comment by ZyX-I** [comment]
Latest variant with populating &runtimepath from various XDG_ variables is 280c5365d5128a4b1e6419b2ead6c4fe25c5f5ed.

@fmoralesc Pathogen is not a plugin manager.

File: src/nvim/os/env.c [hunk 14]
```diff
@@ -427,6 +428,37 @@ char *vim_getenv(const char *name)
...
+    } else if (strcmp(name, "XDG_CONFIG_HOME") == 0) {
+      home_suffix = "/.config/";
+    } else if (strcmp(name, "XDG_CACHE_HOME") == 0) {
+      home_suffix = "/.cache/";
+    } else if (strcmp(name, "XDG_RUNTIME_DIR") == 0) {
+      return xstrdup("/tmp/");  // Arbitrary value, any idea?
+    } else if (strcmp(name, "XDG_CONFIG_DIRS") == 0) {
+      return xstrdup("/etc/xdg");
+    } else if (strcmp(name, "XDG_DATA_DIRS") == 0) {
+      return xstrdup("/usr/local/share/:/usr/share/");
```
Comment (ZyX-I) [inline src/nvim/os/env.c]: Anybody knows why here are trailing slashes and above are not? (I have copied defaults from the specification, so don’t say “because you wrote that”.)

**Comment by fmoralesc** [comment]
> pathogen is not a plugin manager

@ZyX-I I know, and I knew someone would comment on it. But it falls within the category I wanted to test, and it was simpler to describe it like this.

**Comment by ZyX-I** [comment]
About plugins not fine with `~/.config/nvim`: I have saw some and they are neither fine with

1. `~/.nvim`.
2. Vim installed on Windows.
3. According to the documentation: Vim installed on Amiga, OS/2, Macintosh, RISC-OS and VMS (literally everything, but linux).

: because they hardcode `~/.vim`.

All plugin managers listed as not fine with the change allow configuring directories, *correctly written* plugins should either iterate over `&runtimepath` or use `expand('<sfile>:…')`. Note: VAM by default *first* checks whether it can install plugins to the location where it is installed on its own, so if user is not using system-wide VAM and also did not configure it to use some specific directory it is completely OK to just do `mv --no-target-directory ~/.nvim ~/.config/nvim`.

File: src/nvim/os/env.c (hunk 12 above)
Comment (Yamakaky) [inline src/nvim/os/env.c]: https://github.com/Yamakaky/neovim/commit/a1429b850c634081579176f524ec8fc2eec46e88

File: src/nvim/os/env.c (hunk 13 above)
Comment (Yamakaky) [inline src/nvim/os/env.c]: https://github.com/Yamakaky/neovim/commit/a1429b850c634081579176f524ec8fc2eec46e88

File: src/nvim/os/unix_defs.h (hunk 9 above)
Comment (Pyrohh) [inline src/nvim/os/unix_defs.h]: Oh wow, that's interesting. Thanks.

**Comment by ZyX-I** [comment]
Commit bf4a342e77c659f71eb5f16106b22337480e5a54 adds escaping commas.

**Comment by ZyX-I** [comment]
I am wondering whether anybody is actually supporting using `XDG_*_DIRS` for default &runtimepath.

File: src/nvim/os/unix_defs.h [hunk 15]
```diff
@@ -28,46 +32,51 @@
...
-# define DFLT_RUNTIMEPATH "~/.nvim," RUNTIME_GLOBAL ",$VIMRUNTIME," \
-  RUNTIME_GLOBAL "/after,~/.nvim/after"
+# define DFLT_RUNTIMEPATH NVIM_CONF_DIR "," RUNTIME_GLOBAL ",$VIMRUNTIME," \
+                          RUNTIME_GLOBAL "/after," NVIM_CONF_DIR "/after"
 #else
 # define DFLT_RUNTIMEPATH \
-  "~/.nvim,$VIM/vimfiles,$VIMRUNTIME,$VIM/vimfiles/after,~/.nvim/after"
+  NVIM_CONF_DIR ",$VIM/vimfiles,$VIMRUNTIME,$VIM/vimfiles/after,"\
+  NVIM_CONF_DIR "/after"
 #endif
```
Comment (justinmk) [inline src/nvim/os/unix_defs.h]: This entire `#else` block can be deleted, as well as all mentions of `RUNTIME_GLOBAL` in the source.

File: src/nvim/os/unix_defs.h (hunk 9 above)
Comment (ZyX-I) [inline src/nvim/os/unix_defs.h]: `//` does not look clean. I would avoid it when avoiding it costs nothing: just remove trailing slash here because in other places you already use leading slash.

File: src/nvim/os/unix_defs.h [hunk 16]
```diff
@@ -28,46 +32,51 @@
 #ifndef SYNTAX_FNAME
 # define SYNTAX_FNAME   "$VIMRUNTIME/syntax/%s.vim"
 #endif
+// Not used anymore ?
```
Comment (justinmk) [inline src/nvim/os/unix_defs.h]: `USR_EXRC_FILE`, `USR_VIMRC_FILE2`, `EXRC_FILE`,  can be removed. But not `VIMRC_FILE`.

File: src/nvim/os/unix_defs.h [hunk 17]
```diff
@@ -28,46 +32,51 @@
...
 #ifndef USR_VIMRC_FILE2
 # define USR_VIMRC_FILE2     "~/.nvim/nvimrc"
 #endif
+// Not used anymore ?
 #ifndef EXRC_FILE
 # define EXRC_FILE      ".exrc"
 #endif
+// Not used anymore ?
 #ifndef VIMRC_FILE
 # define VIMRC_FILE     ".nvimrc"
```
Comment (justinmk) [inline src/nvim/os/unix_defs.h]: This is for `'exrc'` feature. Removing this is out of scope, will be a long discussion.

**Comment by Pyrohh** [comment]
@ZyX-I 

> I am wondering whether anybody is actually supporting using XDG_*_DIRS for default &runtimepath.

I'm a little confused by your phrasing: are you saying that people wouldn't use this? Up until recently I wasn't aware of a use case for XDG_*_DIRS, but now...

I set `XDG_DATA_DIRS=/usr/local/share:/home/michael/.racket/6.2/share` so i3-dmenu-desktop can pick up `/home/michael/.racket/6.2/share/applications/drracket.desktop`. I do this because `XDG_DATA_HOME` is already set to `~/.local/share`, and IIRC it can only contain a single directory path.

Admittedly, I can't think of a practical use case specific to nvim, since instead of dealing with environment variables, a user could just append to `&rtp` directly. Consider it food for thought I guess.

File: src/nvim/os/env.c (hunk 13 above)
Comment (justinmk) [inline src/nvim/os/env.c]: Why not use `concat_str()` or `vim_strcat()` from `nvim/strings.h`?

**Comment by justinmk** [comment]
As mentioned in https://github.com/neovim/neovim/pull/3125#discussion_r36040649 , after the mechanics are decided here, this PR will require:

1. Documentation. All references to ~/.vim and ~/.nvim should be found and corrected.
2. Unit tests for `vim_getenv()`
3. Functional tests that exercise the default values _and_ test that user overrides are respected.

We can probably move forward without deciding Windows defaults at the moment, but we'll need to address that shortly thereafter.

**Comment by ZyX-I** [comment]
> I'm a little confused by your phrasing: are you saying that people wouldn't use this? Up until recently I wasn't aware of a use case for XDG_*_DIRS, but now...

I am just asking whether this decision is supported by the developers.

> Admittedly, I can't think of a practical use case specific to nvim, since instead of dealing with environment variables, a user could just append to &rtp directly. Consider it food for thought I guess.

This change is here to support XDG base directory specification. XDG_CONFIG_DIRS is there because specification requires to search configuration there. XDG_DATA_HOME+XDG_DATA_DIRS are there because /usr/share is a data directory and /usr/share/nvim is being used for plugin files (so we are already using directory from XDG_DATA_DIRS for our purposes).

**Comment by ZyX-I** [comment]
> I set XDG_DATA_DIRS=/usr/local/share:/home/michael/.racket/6.2/share so i3-dmenu-desktop can pick up /home/michael/.racket/6.2/share/applications/drracket.desktop. I do this because XDG_DATA_HOME is already set to ~/.local/share, and IIRC it can only contain a single directory path.

It is rather strange setting:

1. `/usr/share` is missing. It is from defaults and this will break whatever software installed by your system package manager which uses XDG_DATA_DIRS in place of hardcoding prefix=/usr somewhere. If there is such software.
2. `~/…` directory is after `/usr/local`. I would expect any user-local directory have greater priority.

**Comment by Yamakaky** [comment]
BTW, do we want to support XDG on windows (with other defaults) ? I think it may simplify the code if we do that.

File: src/nvim/os/unix_defs.h (hunk 16 above)
Comment (Yamakaky) [inline src/nvim/os/unix_defs.h]: WIP. Do we also drop EXINIT ?

File: src/nvim/os/unix_defs.h (hunk 17 above)
Comment (Yamakaky) [inline src/nvim/os/unix_defs.h]: ... So what about EXRC_FILE you told me to drop ?

Comment (justinmk) [inline src/nvim/os/unix_defs.h]: Not related.

File: src/nvim/os/unix_defs.h (hunk 16 above)
Comment (justinmk) [inline src/nvim/os/unix_defs.h]: No.

**Comment by justinmk** [comment]
Yes.

File: src/nvim/os/unix_defs.h (hunk 17 above)
Comment (Yamakaky) [inline src/nvim/os/unix_defs.h]: Oh, sorry, I thought 'exrc' was related to .exrc

**Comment by Pyrohh** [comment]
> /usr/share is missing. It is from defaults and this will break whatever software installed by your system package manager which uses XDG_DATA_DIRS in place of hardcoding prefix=/usr somewhere. If there is such software

Indeed, only if there is such software. I'm using OpenBSD, so all 3rd-party software is in `/usr/local`, and there is no `/usr/share/applications`, so it works fine for me.

> ~/… directory is after /usr/local. I would expect any user-local directory have greater priority

Perhaps that's a good idea, although it hasn't been an issue in practice.

**Comment by jck** [comment]
Rather than relying on macros, I think we should do something like this: https://github.com/jck/neovim/commit/fb4a513ba2776e845b7ef25032f33a736130b3a7

**Comment by Pyrohh** [comment]
closing in favor of #3198
